### ISSUE TITLE
Benchmark and optimization of `nerdctl images`

### ISSUE THREAD

**Issue by AkihiroSuda** [containerd/nerdctl#809]
>`nerdctl images` might be slower than before (which has been already slow).
>
>We should do some benchmark and probably have `--quick` (`--fast`?, idk) flag to skip slow size calculation

_Originally posted by @AkihiroSuda in https://github.com/containerd/nerdctl/issues/789#issuecomment-1038903152_

**Comment by ningmingxiao** [comment]
I think  nerdctl image  -a or -o wide  should show blobsize,at default it is not necessary to show it.

**Comment by fahedouch** [comment]
benchmark 

Only `Unpacked Size`:

```
ntimes -n 10 _output/nerdctl images

REPOSITORY    TAG       IMAGE ID        CREATED           PLATFORM       SIZE         BLOB SIZE
alpine        latest    21a3deaa0d32    24 hours ago      linux/amd64    5.9 MiB      0.0 B
debian        latest    fb45fd4e25ab    14 minutes ago    linux/amd64    134.5 MiB    0.0 B
ubuntu        latest    669e010b58ba    13 minutes ago    linux/amd64    77.9 MiB     0.0 B

real average: 164.16197ms, max: 185.0302ms, min: 137.0472ms, std dev: 16.171173ms
real 99 percentile: 185.0302ms, 95 percentile: 185.0302ms, 50 percentile: 165.57285ms
user average: 54.0097ms, max: 64.725ms, min: 41.953ms, std dev: 7.904296ms
sys  average: 99.1182ms, max: 119.68ms, min: 79.722ms, std dev: 13.591908ms
flaky: 0%
```

Only  `Blob Size`:

```
REPOSITORY    TAG       IMAGE ID        CREATED           PLATFORM       SIZE     BLOB SIZE
alpine        latest    21a3deaa0d32    24 hours ago      linux/amd64    0.0 B    2.7 MiB
debian        latest    fb45fd4e25ab    17 minutes ago    linux/amd64    0.0 B    52.4 MiB
ubuntu        latest    669e010b58ba    16 minutes ago    linux/amd64    0.0 B    27.2 MiB

real average: 157.88058ms, max: 236.3843ms, min: 142.0371ms, std dev: 28.001696ms
real 99 percentile: 236.3843ms, 95 percentile: 236.3843ms, 50 percentile: 149.58565ms
user average: 53.0145ms, max: 79.649ms, min: 43.332ms, std dev: 11.046308ms
sys  average: 86.4968ms, max: 100.921ms, min: 72.189ms, std dev: 8.907955ms
flaky: 0%
```

Both : 

```
REPOSITORY    TAG       IMAGE ID        CREATED           PLATFORM       SIZE         BLOB SIZE
alpine        latest    21a3deaa0d32    24 hours ago      linux/amd64    5.9 MiB      2.7 MiB
debian        latest    fb45fd4e25ab    20 minutes ago    linux/amd64    134.5 MiB    52.4 MiB
ubuntu        latest    669e010b58ba    19 minutes ago    linux/amd64    77.9 MiB     27.2 MiB

real average: 184.36679ms, max: 217.7352ms, min: 170.1355ms, std dev: 15.003351ms
real 99 percentile: 217.7352ms, 95 percentile: 217.7352ms, 50 percentile: 174.8397ms
user average: 65.1658ms, max: 82.53ms, min: 52.764ms, std dev: 8.821194ms
sys  average: 106.5241ms, max: 112.316ms, min: 92.911ms, std dev: 5.815795ms
flaky: 0%
```

the difference between the 3 average is negligible I think It is useless to make a `--quick` flag

**Comment by AkihiroSuda** [comment]
Thanks, but we need to benchmark with more than 100 images, I guess

**Comment by fahedouch** [comment]
@AkihiroSuda I agree. But 100 is a huge number. Do you have a suggestion to get this number of image in one environnement 😅

**Comment by apostasie** [comment]
Reviving this issue, and sharing notes here:

I was working on adding caching to the snapshotter for commands like `images` (so that we do not make several `Stat` and `Usage` calls for the same objects), and started looking into image list overall.

Our current implementation of `image list` first calls on `images.List`.

Then, for each image, we call `image.Platforms()`, which:
- reads the manifest blob (again)
- reads the blob of every children of the target that are relevant - if the image is a straight manifest, it will read the config

Then, for each image and each architecture, we call on `images.Check`, which:
- reads the manifest blob (again)
- possibly reads the config blob as well
- then `ReaderAt` on every "required" resource (eg: any linked manifest)

`ReaderAt` does:
- os.Stat
- os.Open

Then we call `containerd.NewImageWithPlatform`, then `Config()` on the resulting image, which will read the config blob again.

Then we call `RootFS` for the size calculation, which again will read the config blob (twice).

Then we call `image.Size` for the blob size calculation, which will again read the blob of every children of the target.

That is a lot of duplicate filesystem calls.

A quick auditd test confirms the above and reports about 60 filesystem hits for Alpine here (alpine contains one layer and 8 architectures).
Whether or not this is entirely accurate, or my reading of the code is 100% right, this is still a lot of filesystem access for *one single image*.

Suggestion:
- we could bypass entirely all the containerd helpers methods and use the content store directly to avoid all of that -  containerd helpers are great for single objects where io does not matter much, but  just not good for the task at hand and clearly problematic at scale
- we change the logic so that we stop repeating these operations on the same image for every platform, and read things *once* wherever possible

That should bring down the number of filesystem access significantly.

Here are some highly not scientific numbers against main and against my current work in progress:

For 500 identical, single platform images (tagged from `apostasie/test`), time will give:
- currently:
  - ~1.5 seconds
- with rewrite:
  - ~0.4 seconds

For 500 identical, multi platform images (all tagged from `alpine:latest`, just the current platform):
- currently:
  - ~3.3 seconds
- with rewrite:
  - ~0.8 seconds

For 500 identical, multi truly platform images (all tagged from `alpine:latest`, --all-platform):
- currently:
  - ~12 seconds
- with rewrite:
  - ~2.9 seconds

Using the image list from above, with code instrumentation, printImage execution time gets divided by 3 (from about 1 second to 0.3 second), although `time` only shows about 20% improvement overall.

The PR is not ready to be merged and more a proof of concept, but would like to know what you folks think.

Cheers.

**Comment by apostasie** [comment]
PR is here: #3029

**Comment by apostasie** [comment]
On today's main (873a08791f570bafb91dc10f99e2b9bf45c7e6f9), with the list of images provided above (84 images currently), on a lima setup / M1 Max:

```
real average: 408.938725ms, max: 430.797757ms, min: 391.571417ms, std dev: 14.795711ms
real 99 percentile: 430.797757ms, 95 percentile: 430.797757ms, 50 percentile: 398.672759ms
user average: 167.9466ms, max: 201.888ms, min: 133.506ms, std dev: 25.245575ms
sys  average: 151.5746ms, max: 173.962ms, min: 110.628ms, std dev: 22.000527ms
flaky: 0%
```
